crates/egui/src/atomics/atomic_layout.rs

```diff
@@ -0,0 +1,392 @@
+            // The `shrink` item gets the remaining space
+            let shrunk_size = Vec2::new(
+                available_inner_size.x - desired_width,
+                available_inner_size.y,
+            );
+            if item.grow {
+                grow_count += 1;
+            }
+
+            let sized = item.into_sized(ui, shrunk_size, max_font_size, Some(wrap_mode));
+            let size = sized.size;
+
+            desired_width += size.x;
+            preferred_width += sized.preferred_size.x;
+
+            height = height.at_least(size.y);
+            preferred_height = preferred_height.at_least(sized.preferred_size.y);
+
+            sized_items.insert(index, sized);
+        }
+
+        let margin = frame.total_margin();
+        let desired_size = Vec2::new(desired_width, height);
+        let frame_size = (desired_size + margin.sum()).at_least(min_size);
+
+        let (_, rect) = ui.allocate_space(frame_size);
+        let mut response = ui.interact(rect, id, sense);
+
+        response.intrinsic_size =
+            Some((Vec2::new(preferred_width, preferred_height) + margin.sum()).at_least(min_size));
+
+        AllocatedAtomicLayout {
+            sized_atomics: sized_items,
+            frame,
+            fallback_text_color,
+            response,
+            grow_count,
+            desired_size,
+            align2,
+            gap,
+        }
+    }
+}
+
+/// Instructions for painting an [`AtomicLayout`].
+#[derive(Clone, Debug)]
+pub struct AllocatedAtomicLayout<'a> {
+    pub sized_atomics: Vec<SizedAtomic<'a>>,
+    pub frame: Frame,
+    pub fallback_text_color: Color32,
+    pub response: Response,
+    grow_count: usize,
+    // The size of the inner content, before any growing.
+    desired_size: Vec2,
+    align2: Align2,
+    gap: f32,
+}
+
+impl<'a> AllocatedAtomicLayout<'a> {
+    /// Paint the [`Frame`] and individual [`Atomic`]s.
+    pub fn paint(self, ui: &Ui) -> AtomicLayoutResponse {
+        let Self {
+            sized_atomics: sized_items,
+            frame,
+            fallback_text_color,
+            response,
+            grow_count,
+            desired_size,
+            align2,
+            gap,
+        } = self;
+
+        let inner_rect = response.rect - self.frame.total_margin();
+
+        ui.painter().add(frame.paint(inner_rect));
+
+        let width_to_fill = inner_rect.width();
+        let extra_space = f32::max(width_to_fill - desired_size.x, 0.0);
+        let grow_width = f32::max(extra_space / grow_count as f32, 0.0);
+
+        let aligned_rect = if grow_count > 0 {
+            align2.align_size_within_rect(Vec2::new(width_to_fill, desired_size.y), inner_rect)
+        } else {
+            align2.align_size_within_rect(desired_size, inner_rect)
+        };
+
+        let mut cursor = aligned_rect.left();
+
+        let mut response = AtomicLayoutResponse {
+            response,
+            custom_rects: HashMap::new(),
```

Comment:
less code to change if we change the type of `custom_rects` in the future

```suggestion
            custom_rects: Default::default(),
```